### ISSUE TITLE
iOS can't play rtsp

### ISSUE THREAD

**Issue by YangYuZhe** [Bilibili/ijkplayer#536]

**Comment by YangYuZhe** [comment]
Log:
===== custom modules begin =====
register protocol: async
register protocol: ijkhttphook
register protocol: ijkinject
register protocol: ijklongurl
register protocol: ijksegment
register protocol: ijktcphook
register demuxer : ijklivehook
===== custom modules end =====
av_version_info: ff2.8--ijk0.4.1.1--dev0.3.3--rc2
ijkmp_set_inject_opaque(0x7fc5a1e53770)
ijkmp_set_inject_opaque()=void
2015-10-12 19:14:58.327 IJKMediaDemo[48842:2653032] OK setup GL
ijkmp_ios_set_view(glView=0x7fc5a1cc7ae0)
ijkmp_ios_set_view(glView=0x7fc5a1cc7ae0)=void
ijkmp_set_data_source(url="rtsp://218.204.223.237:554/live/1/0547424F573B085C/gsfp90ef4k0a6iap.sdp")
ijkmp_set_data_source(url="rtsp://218.204.223.237:554/live/1/0547424F573B085C/gsfp90ef4k0a6iap.sdp")=0
ijkmp_prepare_async()
remove 'timeout' option for rtmp.
===== versions =====
FFmpeg       : ff2.8--ijk0.4.1.1--dev0.3.3--rc2
libavutil    : 54.31.100
libavcodec   : 56.60.100
libavformat  : 56.40.101
libswscale   : 3.1.101
libswresample: 1.2.101
===== options =====
player-opts : start-on-prepared            = 1
player-opts : overlay-format               = fcc-i420
player-opts : max-fps                      = 30
player-opts : framedrop                    = 0
player-opts : videotoolbox-max-frame-width = 960
player-opts : videotoolbox                 = 0
player-opts : video-pictq-size             = 3
format-opts : ijkinject-opaque             = 140486801438576
format-opts : user-agent                   = ijkplayer
format-opts : auto_convert                 = 0
format-opts : safe                         = 0
format-opts : reconnect                    = 1
codec-opts  : skip_frame                   = 8
codec-opts  : skip_loop_filter             = 48
===================
ijkmp_prepare_async()=0
remove 'timeout' option for rtmp.
ijkmp_start()
ijkmp_start()=-3
SDP:
v=0
o=- 1 1 IN IP4 0.0.0.0
s=RTSP Session
i=Jabsco Stream(JCO-jy9757acx1eve7nm-692af75ce5c3e76b8b457e2c7248f13b)
c=IN IP4 218.204.223.237
t=0 0
a=range:npt=now-
a=control:*
m=video 0 RTP/AVP 96
b=AS:120
a=rtpmap:96 MP4V-ES/90000
a=fmtp:96 profile-level-id=1;config=000001b001000001b5090000010100000120008449c428582120a21f;decode_buf=115200
a=framerate:10
a=framesize:96 352-288
a=control:trackID=1

video codec set to: (null)
end receive buffer size reported is 65536
end receive buffer size reported is 65536
hello state=0
Option ijkinject-opaque not found.
parser not found for codec mpeg4, packets or times may be invalid.
parser not found for codec mpeg4, packets or times may be invalid.
Non-increasing DTS in stream 0: packet 2 with DTS 0, packet 3 with DTS 0
Non-increasing DTS in stream 0: packet 3 with DTS 0, packet 4 with DTS 0
Non-increasing DTS in stream 0: packet 4 with DTS 0, packet 5 with DTS 0
Non-increasing DTS in stream 0: packet 5 with DTS 0, packet 6 with DTS 0
Non-increasing DTS in stream 0: packet 6 with DTS 0, packet 7 with DTS 0
Non-increasing DTS in stream 0: packet 7 with DTS 0, packet 8 with DTS 0
Non-increasing DTS in stream 0: packet 8 with DTS 0, packet 9 with DTS 0
Non-increasing DTS in stream 0: packet 9 with DTS 0, packet 10 with DTS 0
Non-increasing DTS in stream 0: packet 10 with DTS 0, packet 11 with DTS 0
Non-increasing DTS in stream 0: packet 11 with DTS 0, packet 12 with DTS 0
Non-increasing DTS in stream 0: packet 12 with DTS 0, packet 13 with DTS 0
Non-increasing DTS in stream 0: packet 13 with DTS 0, packet 14 with DTS 0
Non-increasing DTS in stream 0: packet 14 with DTS 0, packet 15 with DTS 0
Non-increasing DTS in stream 0: packet 15 with DTS 0, packet 16 with DTS 0
Non-increasing DTS in stream 0: packet 16 with DTS 0, packet 17 with DTS 0
Non-increasing DTS in stream 0: packet 17 with DTS 0, packet 18 with DTS 0
Non-increasing DTS in stream 0: packet 18 with DTS 0, packet 19 with DTS 0
Non-increasing DTS in stream 0: packet 19 with DTS 0, packet 20 with DTS 0
Non-increasing DTS in stream 0: packet 20 with DTS 0, packet 21 with DTS 0
Non-increasing DTS in stream 0: packet 21 with DTS 0, packet 22 with DTS 0
Non-increasing DTS in stream 0: packet 22 with DTS 0, packet 23 with DTS 0
Non-increasing DTS in stream 0: packet 23 with DTS 0, packet 24 with DTS 0
Non-increasing DTS in stream 0: packet 24 with DTS 0, packet 25 with DTS 0
Non-increasing DTS in stream 0: packet 25 with DTS 0, packet 26 with DTS 0
Non-increasing DTS in stream 0: packet 26 with DTS 0, packet 27 with DTS 0
Non-increasing DTS in stream 0: packet 27 with DTS 0, packet 28 with DTS 0
Non-increasing DTS in stream 0: packet 42 with DTS 19484, packet 43 with DTS 19484
Non-increasing DTS in stream 0: packet 43 with DTS 19484, packet 44 with DTS 19484
Non-increasing DTS in stream 0: packet 44 with DTS 19484, packet 45 with DTS 19484
Non-increasing DTS in stream 0: packet 45 with DTS 19484, packet 46 with DTS 19484
Non-increasing DTS in stream 0: packet 46 with DTS 19484, packet 47 with DTS 19484
Non-increasing DTS in stream 0: packet 47 with DTS 19484, packet 48 with DTS 19484
Non-increasing DTS in stream 0: packet 48 with DTS 19484, packet 49 with DTS 19484
Non-increasing DTS in stream 0: packet 49 with DTS 19484, packet 50 with DTS 19484
Non-increasing DTS in stream 0: packet 50 with DTS 19484, packet 51 with DTS 19484
Non-increasing DTS in stream 0: packet 51 with DTS 19484, packet 52 with DTS 19484
Non-increasing DTS in stream 0: packet 52 with DTS 19484, packet 53 with DTS 19484
av_version_info: ff2.8--ijk0.4.1.1--dev0.3.3--rc2
ijkmp_set_inject_opaque(0x7fc5a4a04790)
ijkmp_set_inject_opaque()=void
2015-10-12 19:14:58.885 IJKMediaDemo[48842:2653032] OK setup GL
ijkmp_ios_set_view(glView=0x7fc5a4a04f40)
ijkmp_ios_set_view(glView=0x7fc5a4a04f40)=void
2015-10-12 19:14:58.886 IJKMediaDemo[48842:2653032] Warning: Attempt to present <IJKVideoViewController: 0x7fc5a1fb1830> on <IJKDemoInputURLViewController: 0x7fc5a1e4f120> whose view is not in the window hierarchy!
[rtsp @ 0x7fc5a2040e00] Non-increasing DTS in stream 0: packet 53 with DTS 19484, packet 54 with DTS 19484
[rtsp @ 0x7fc5a2040e00] Non-increasing DTS in stream 0: packet 54 with DTS 19484, packet 55 with DTS 19484
[rtsp @ 0x7fc5a2040e00] Non-increasing DTS in stream 0: packet 55 with DTS 19484, packet 56 with DTS 19484

**Comment by Zard1096** [comment]
```parser not found for codec mpeg4, packets or times may be invalid.```
Probably you need add decoder with mpeg4 in config

**Comment by YangYuZhe** [comment]
very appreciated!
And do you know how to fetch CVImageBuffer object?I see that in this https://github.com/Bilibili/ijkplayer/blob/master/ios/IJKMediaPlayer/IJKMediaPlayer/ijkmedia/ijksdl/ios/ijksdl_vout_overlay_videotoolbox.m#L148
But i don't know how to manage it in detail...I haven't ever use C function in OC.m
Could you please mention a little for me,thanks a lot

**Comment by Zard1096** [comment]
I have no idea about how to manage it in detail.What exactly you want to do?
If just fetch the CVPixelBufferRef you can do it after getting object in https://github.com/Bilibili/ijkplayer/blob/master/ios/IJKMediaPlayer/IJKMediaPlayer/ijkmedia/ijksdl/ios/IJKSDLGLRenderNV12.m#L117
CVImageBufferRef is a synonym for a CVPixelBufferRef,see Apple Docs
```typedef CVImageBufferRef CVPixelBufferRef;```

**Comment by YangYuZhe** [comment]
right right!I'm just want to fetch CVPixelBufferRef to create texture for OpenGL stuff!!
Then,should I do these for that?:IJKFFMoviePlayerController -> _mediaPlayer(IJKMediaPlayer *)-> ffplayer-> bout ->SDL_VoutOverlay  And send it to SDL_VoutOverlayVideoToolBox_GetCVPixelBufferRef(arg)?
VERY APPRECIATE! I'm confused by this issue for many many days!

**Comment by Zard1096** [comment]
You are free to modify what you want.

**Comment by YangYuZhe** [comment]
Can't thank you anymore XD

**Comment by aldirc** [comment]
hello!
`parser not found for codec mpeg4, packets or times may be invalid.`
how to figure it?

**Comment by zaferaydn** [comment]
Can someone tell me how to enable **mpeg4** decoders? What do i need to add in **module.sh** file?

**Comment by raymond1012** [comment]
search issue

2017-05-08 22:02 GMT+08:00 Zafer <notifications@github.com>:

**Comment by zaferaydn** [comment]
@raymond1012  thank you raymond. I searched and i found a similar issue. From there I investigate how to enable **mpeg4**. In general, in order to add a needed decoder,demuxer,parser or protocol:

_In ffmpeg directory, run_
```
./configure --list-decoders
./configure --list-demuxers
./configure --list-parsers
./configure --list-protocols
```

These commands will show which options are available. I just chose the options which were related with **mpeg4**. Added these to **modules.sh** file, rebuild the source and successed.

Thanks.
